iroh/src/magicsock.rs

```diff
@@ -1561,7 +1633,50 @@ enum DiscoBoxError {
     Parse(anyhow::Error),
 }
 
-type RelayRecvResult = Result<(PublicKey, quinn_udp::RecvMeta, Bytes), io::Error>;
+/// Channel for [`MagicSock::poll_recv_relay`] to receive QUIC datagrams from relays.
+///
+/// This channel contains both the send and receive sides.  The receive side implements the
+/// poll interface with required wakers.
+#[derive(Debug)]
+struct RelayRecvChannel {
+    sender: mpsc::Sender<RelayReadResult>,
+    receiver: parking_lot::Mutex<mpsc::Receiver<RelayReadResult>>,
+    waker: parking_lot::Mutex<Option<Waker>>,
+}
+
+impl RelayRecvChannel {
+    fn new() -> Self {
+        let (tx, rx) = mpsc::channel(128);
+        Self {
+            sender: tx,
+            receiver: parking_lot::Mutex::new(rx),
+            waker: parking_lot::Mutex::new(None),
+        }
+    }
+
+    fn poll_recv(&self, cx: &mut Context) -> Poll<RelayReadResult> {
+        match self.receiver.lock().try_recv() {
+            Ok(item) => Poll::Ready(item),
+            Err(mpsc::error::TryRecvError::Empty) => {
+                self.waker.lock().replace(cx.waker().clone());
+                Poll::Pending
```

Comment:
Somebody who knows more about rust syntax/semantics than I do:
Does this mean that both locks `receiver` and `waker` are held at the same time, or does this mean that `receiver` is locked, then dropped, and then `waker` is locked, then dropped?

If the latter, then there's a potential for a race. Not a biggie, just re-check `receiver` after registering the waker, or make sure to keep around the `receiver` lock, but still.

Reply:
My expectation is that the `receiver` is only locked while executing the `try_recv` call and unlocked immediately after, before the match is executed.  My understanding is that the guard is a temporary variable that gets dropped at the end of the expression.

Same with the `waker` lock, I expect it to be released at the end of that expression.

https://smallcultfollowing.com/babysteps/blog/2023/03/15/temporary-lifetimes/
and
https://blog.m-ou.se/super-let/
are good discussions of this topic.

**But**!  That doesn't mean I got it wrong.  If we're unsure I could re-write this to be explicit about where the guards are dropped?

Reply:
Good catch!  I believe this bug was there already in the old implementation too.  :scream:

Reply:
eeeh.  This commit broke things :(